packages/federation/src/supergraph.ts

```diff
@@ -804,12 +826,75 @@ export function getSubschemasFromSupergraphSdl({
         return res;
       };
     }
+    function visitToReplaceTypeNames(data: any, nameMap: Map<string, string>): any {
+      if (data != null && typeof data === 'object') {
+        if (Array.isArray(data)) {
+          return data.map(item => visitToReplaceTypeNames(item, nameMap));
+        }
+        const newData = {};
+        for (const key in data) {
+          if (key === '__typename') {
+            const newType = nameMap.get(data[key]) || data[key];
+            newData[key] = newType;
+          } else {
+            newData[key] = visitToReplaceTypeNames(data[key], nameMap);
+          }
+        }
+        return newData;
+      }
+      return data;
+    }
     subschemaMap.set(subgraphName, {
       name: subgraphName,
       schema,
       executor,
       merge: mergeConfig,
       batch,
+      transforms: [
+        {
+          transformRequest(request, delegationContext, transformationContext) {
+            delegationContext.args = visitToReplaceTypeNames(
+              delegationContext.args,
+              fakeTypesIfaceMap,
+            );
+            return {
+              ...request,
+              document: visit(request.document, {
+                [Kind.INLINE_FRAGMENT](node) {
```

Comment:
Just out of curiosity, why use this instead of?
```suggestion
                InlineFragment(node) {
```

Reply:
Cool! Was just wondering.